src/api_types.jl

```diff
@@ -0,0 +1,310 @@
+## C types
+const C_time_t = Int
+
+## HDF5 types and constants
+const haddr_t  = UInt64
+const herr_t   = Cint
+const hid_t    = Int64
+const hsize_t  = UInt64
+const hssize_t = Int64
+const htri_t   = Cint   # pseudo-boolean (negative if error)
+
+# For VLEN
+struct hvl_t
+    len::Csize_t
+    p::Ptr{Cvoid}
+end
+const HVL_SIZE = sizeof(hvl_t)
+
+# Object reference types
+struct HDF5ReferenceObj
+    r::UInt64 # Size must be H5R_OBJ_REF_BUF_SIZE
+end
+const HDF5ReferenceObj_NULL = HDF5ReferenceObj(UInt64(0))
+
+# For group information
+struct H5G_info_t
+    storage_type::Cint # enum H5G_storage_type_t
+    nlinks::hsize_t
+    max_corder::Int64
+    mounted::Cuint
+end
+
+# For objects
+struct H5_ih_info_t
+    index_size::hsize_t
+    heap_size::hsize_t
+end
+struct H5O_info_t # version 1 type H5O_info1_t
+    fileno::Cuint
+    addr::haddr_t
+    otype::Cint # enum H5O_type_t
+    rc::Cuint
+    atime::C_time_t
+    mtime::C_time_t
+    ctime::C_time_t
+    btime::C_time_t
+    num_attrs::hsize_t
+    #{ inlined struct H5O_hdr_info_t named type
```

Comment:
I remember looking at this before and thinking that this inlining seems to directly contradict the [documentation](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Struct-Type-Correspondences). I know you didn't do this, but I'm just raising the point now since you added the nice comments and made me think of it again. Do we know if this actually works or if the alignment is messed up?

Reply:
This shouldn't be a blocker for merging or anything since this PR doesn't change anything. It just made me think of this question that I thought of but didn't have the motivation to figure out last time I saw it.

Reply:
I think that should work — the structure has size and alignment equal to the members, so inlining them shouldn't change that. As a concrete check:
```julia
julia> using HDF5

julia> Int(fieldoffset(HDF5.H5O_info_t, fieldcount(HDF5.H5O_info_t)))
144
```
which can be compared against the `offsetof(H5O_info1_t, meta_size.attr);` computed in C: [godbolt](https://godbolt.org/z/jKsPfs) which calculates it as 144 as well.

Reply:
Ah, just read the link you had to the docs. Maybe the saving grace here is that the sum of the inlined members' sizes correspond to an alignment offset already, so there's no padding to worry about?

Reply:
I'm not opposed to making the fields a separate struct to follow the C API, though.

Reply:
> I'm not opposed to making the fields a separate struct to follow the C API, though.

Probably  a better idea. 

Looking at the docs, I don't think there's an alignment issue here. The result should be the same if we defined the fields in a separate (immutable) struct.  @kleinhenz was there something specific that lead you to believe there was an alignment issue to begin with?

Reply:
No I haven't noticed any problems, although I think this part of the code is not exercised in the tests much. I just noticed that this does what the julia documentation makes a point of saying not to do, but I am definitely not an expert on this kind of thing even in c.

Reply:
The documentation reads wrong to me?

This doesn't make sense to me

>  When mirroring a struct used by-value inside another struct in C, it is imperative that you do not attempt to manually copy the fields over, as this will not preserve the correct field alignment.

from the docs at https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Struct-Type-Correspondences

To me this reads as telling us that we should *not*  mirror  
```c
typedef struct {
    int a;
    double b;
} A;

typedef struct {
    A a;
    A b;
    int c;
} B;
```

But AFAIK it's perfectly correct to mirror this in Julia as follows:
```julia
struct A
    a::Cint
    b::Cdouble
end

struct B
    a::A
    b::A
    c::Cint
end

## or even

struct B
    a1::Cint
    b1::Cdouble
    a2::Cint
    b2::Cdouble
    c::Cint
end
```

and retain correct alignment

Reply:
For reference I've also been referred to : https://github.com/analytech-solutions/CBinding.jl   and https://github.com/JuliaLang/julia/issues/33120 regarding these issues.

Reply:
Yeah, I came across `CBinding.jl` and `CBindingGen.jl` yesterday. My quick/inexpert thought is that it's maybe too heavy for what we want here — it generated a _large_ file of a bunch of macro definitions that took quite a while to precompile and load. Doing it manually seems to give much shorter load time.

That said, I've got a very rough use of `Clang.jl`'s generator running — not well enough that it can actually auto-generate the bindings for us, but enough that it can be a useful point of comparison (and very fast way to replace all the list of consts with enums if desired).